### DISCUSSION TITLE
Best practices for handling dynamic aspect ratios and resizing?

### DISCUSSION THREAD

**Discussion by Joshua-Beatty** [excaliburjs/Excalibur#2269] [Q&A]
Hi All,

I have just stumbled upon Excalibur and it looks incredibly promising and interesting. I am in the process of trying to build a small scale game with the framework to get a feel for it. My end goal would be to have a game that could run on mobile, and any web browser. Looking at the docs [here](https://excaliburjs.com/docs/screens) I see a few different scaling methods. Ideal I would have like a 16:9 "safe area"(This is how i have generally seen it talked about in other engines) like [DisplayMode.FitScreen](https://excaliburjs.com/docs/api/edge/enums/DisplayMode.html#FitScreen), but I would still like to be able to display content outside of this safe area, to give a cohesive look on all aspect ratios, with no black bars. 

Currently my solution is to use [DisplayMode.FillScreen](https://excaliburjs.com/docs/api/edge/enums/DisplayMode.html#FillScreen) and manually calculate a 16 by 9 safe area, but this gets more complicated quickly when resizing happens and all elements need to be re positions. This alone is not a deal breaker, but I want to make sure I am going about this the best way possible, and that I am not missing any easy fixes for this from the documentation.

<hr>

Edit:
Another Idea I have is to something along these lines:
```javascript
const vw = Math.max(document.documentElement.clientWidth || 0, window.innerWidth || 0)
const vh = Math.max(document.documentElement.clientHeight || 0, window.innerHeight || 0)
const game = new Engine({
    width: vw,
    height: vh,
    canvasElementId: "game",
    displayMode: DisplayMode.FitScreen,
});
```
Where the aspect ratio is dynamic upon the initial load but then uses `FitScreen` for that ratio. This gives a good experiences on browsers and mobile devices, and if resized it still acts mostly okay.

<hr>

Edit 2:

Another idea I have:
```javascript
const vw = Math.max( document.documentElement.clientWidth || 0, window.innerWidth || 0 );
const vh = Math.max( document.documentElement.clientHeight || 0, window.innerHeight || 0 );
let game;
if (vw / vh <= 16 / 9) {
  game = new Engine({
    width: vw * 1280 / vw,
    height: vw * 1280 / vw * vh / vw,
    canvasElementId: "game",
    displayMode: DisplayMode.FitScreen,
    backgroundColor: Color.Black,
  });
} else {
  game = new Engine({
    width: vh * 720 / vh * vw / vh,
    height: vh * 720 / vh,
    canvasElementId: "game",
    displayMode: DisplayMode.FitScreen,
    backgroundColor: Color.Black,
  });
}
game.currentScene.camera.x = 0;
game.currentScene.camera.y = 0;
````
So you always have a 1280*720 logical pixels in the center, that you can position all elements absolutely around, but still has the same downside as before of when it is resizes, but it is significantly easier to position the different elements.

**Comment by eonarheim** [comment]
Hi @Joshua-Beatty

Thanks for the kind words!

This is definitely a desirable feature to allow a safe area without black bars. I'll add an issue to look into this being rolled into Excalibur. I'm away from my computer for a while but I'll document it as soon as I can.

Would you want something that would preserve the aspect ratio, but "scale" and center the viewport to hide the black bars? It looks like unity will return a read-only safe area bounding box based on the current screen over scan, does something like that sound like it meets your expectations?
 https://docs.unity3d.com/ScriptReference/Screen-safeArea.html 

Do you have an example of what the most desirable resizing experience might look like to you in your use case?

In the current state I think your solutions are the best that can be done without changing Excalibur. Very cool stuff you've got!

cc @jedeen @kamranayub

**Reply by kamranayub** [reply]
This is a great idea. I know for https://excaliburjs.com/sweep/  I had to do a lot of finicky pixel repositioning when resizing to keep things relative to each other. If you resize the window for that, you can kind of see what we ended up with.

**Reply by Joshua-Beatty** [reply]
The best solution I have worked with in regards to scaling and resizing is [Solar 2D's](https://docs.coronalabs.com/guide/basics/configSettings/index.html) solution. In Solar 2D you define a "content area" in logical pixels, which is always guaranteed to be maximized on screen. Then in addition to having content width, you have viewable width, and safe viewable width. In addition, to make positioning consistent, (0, 0) would always be the top left of the content area by default. 

So to position game elements you position them absolutely with the content area's size in mind, then you can position a background image on viewable width and other UI off of the safe viewable width.

<hr>

For optimal resizing the content area would scale exactly like DisplayMode.FitScreen, in how the content is area maxed, but with additional canvas space to the sides or above.

Then there would be two solutions for resizing for the rest of the content.

Actors(primarily UI actors, as game objects positioned within the content area would always be safe) could listen to a "resize" event, at which point they would reposition themselves based off of the viewable width or safe viewable width.

The other idea would be to allow UI elements to be positioned based off of the edge of the viewable or safe viewable width, at which point their position would be recalculated on their own when a resize event happens, though I think just having the resize event would be enough, as I am not sure how this would be implemented.

**Reply by Joshua-Beatty** [reply]
I went ahead and created [a fork for this](https://github.com/Joshua-Beatty/Excalibur/tree/contentFit) where I could test out implementing this. It works fantastic, but only as long as the camera is set to 0, 0. 

I think the x and y of the camera would need to modified on resize(To keep the current "content area" centered)? I am unsure if a screen/camera offset like this is something that is currently supported in the code base, because I have mainly just been poking around.

**Reply by Joshua-Beatty** [reply]
Using that same fork, and locking the camera to an actor at the center of the content, even with 0,0 at the top left, gives the perfect behavior.

**Reply by Joshua-Beatty** [reply]
Some further testing, it looks like my first implementation actually works correctly, and that the camera just needs to re positioned once when the game is initialized.

**Reply by eonarheim** [reply]
Hi @Joshua-Beatty I think your fork is pretty close to the solution we are looking for! I like the new `DisplayMode.FitContent` name.

If you want to contribute, please make a PR from your fork and we can help sort out any issues you've identified.

**Reply by chrisMYchen** [reply]
Just wanted to comment that I am picking this back up. I used the code from @Joshua-Beatty's fork and it worked perfectly for what I was expecting. Thanks! 

I opened a new discussion related to me trying to test this and develop with a fork here. I'd love to open a PR if I can get some help with this! https://github.com/excaliburjs/Excalibur/discussions/2465
